text/2978-stack_based_vec.md

```diff
@@ -0,0 +1,301 @@
+
+```rust
+// core::collections
+
+pub struct ArrayVec<T, const N: usize> {
+    data: MaybeUninit<[T; N]>,
+    len: usize
+}
+
+impl<T, const N: usize> ArrayVec<T, N> {
+    // Constructors
+
+    pub unsafe fn from_raw_parts(_ptr: *mut T, _len: usize) -> Self;
+
+    #[inline]
+    pub const fn new() -> Self;
+
+    // Methods
+
+    #[inline]
+    pub fn as_mut_ptr(&mut self) -> *mut T;
+
+    #[inline]
+    pub fn as_mut_slice(&mut self) -> &mut [T];
+
+    #[inline]
+    pub fn as_ptr(&self) -> *const T;
+
+    #[inline]
+    pub fn as_slice(&self) -> &[T];
+
+    #[inline]
+    pub const fn capacity(&self) -> usize;
+
+    pub fn clear(&mut self);
+
+    pub fn dedup(&mut self);
+
+    pub fn dedup_by<F>(&mut self, same_bucket: F)
+    where
+        F: FnMut(&mut T, &mut T) -> bool;
+
+    pub fn dedup_by_key<F, K>(&mut self, key: F)
+    where
+        F: FnMut(&mut T) -> K,
+        K: PartialEq<K>;
+
+    pub fn drain<R>(&mut self, range: R)
+    where
+        R: RangeBounds<usize>;
+
+    pub fn extend_from_cloneable_slice(&mut self, other: &[T])
+    where
+        T: Clone; // Panics at run-time
+
+    pub fn extend_from_copyable_slice(&mut self, other: &[T])
+    where
+        T: Copy; // Panics at run-time
+
+    pub fn insert(&mut self, idx: usize, element: T); // Panics at run-time
+
+    #[inline]
+    pub const fn is_empty(&self) -> bool;
+
+    #[inline]
+    pub const fn len(&self) -> usize;
+
+    pub fn pop(&mut self) -> T; // Panics at run-time
+
+    pub fn push(&mut self, element: T); // Panics at run-time
+
+    pub fn remove(&mut self, idx: usize) -> T; // Panics at run-time
+
+    pub fn retain<F>(&mut self, f: F)
+    where
+        F: FnMut(&mut T) -> bool;
+
+    #[inline]
+    pub unsafe fn set_len(&mut self, len: usize);
+
+    pub fn splice<R, I>(&mut self, range: R, replace_with: I)
+    where
+        I: IntoIterator<Item = T>,
+        R: RangeBounds<usize>;
+
+    pub fn split_off(&mut self, at: usize) -> Self;
+
+    pub fn swap_remove(&mut self, idx: usize) -> T;  // Panics at run-time
+
+    pub fn truncate(&mut self, len: usize);
+
+    pub fn try_extend_from_cloneable_slice(&mut self, other: &[T]) -> Result<(), ArrayVecError>
+    where
+        T: Clone;
+
+    pub fn try_extend_from_copyable_slice(&mut self, other: &[T]) -> Result<(), ArrayVecError>
+    where
+        T: Copy;
+
+    pub fn try_insert(&mut self, _idx: usize, element: T) -> Result<(), ArrayVecError>;
+
+    pub fn try_pop(&mut self) -> Result<T, ArrayVecError>;
+
+    pub fn try_push(&mut self, element: T) -> Result<(), ArrayVecError>;
+
+    pub fn try_remove(&mut self, idx: usize) -> Result<T, ArrayVecError>;
+
+    pub fn try_swap_remove(&mut self, idx: usize) -> Result<T, ArrayVecError>;
+}
+
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub enum ArrayVecError {
+    CapacityOverflow,
```

Comment:
```suggestion
    CapacityOverflow(T),
```
Otherwise `element` is lost. Considering that afaict there is no method returning both `CapacityOverflow` and `NoElement` it also makes more sense to split this error into two distinct types.

Reply:
`ArrayVecError` was intended to be a "catch-all" error container but doesn't seem to be necessary at all.

For a better overall coherence, `pub fn push(&mut self, element: T) -> Result<(), ArrayVecError>` is now `pub fn push(&mut self, element: T) -> Result<(), T>`